### PR TITLE
Migrate records to TypeScript

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/nteract/nteract/issues/3592.

This should be rebased and merged after https://github.com/nteract/nteract/pull/3690 is merged.

### CODE REVIEW

**Comment by captainsafia** [comment]
Okie dokie. This is good for review and merge.

**Review by willingc — APPROVED** [review]
Nice @captainsafia!

File: packages/records/src/index.ts [hunk 1]
```diff
@@ -55,7 +53,7 @@ const {
 
 export type Notebook = v4Notebook;
 
-function freezeReviver(k: string, v: JSONType): JSONType {
+function freezeReviver(k: string, v: JSONType): Readonly<JSONType> {
```
Comment (rgbkrk) [inline packages/records/src/index.ts:56]: Smart.

**Review by rgbkrk — APPROVED** [review]
A+. I really like this records API for our core data types (vs. what it's been across commutable + related). Wish we could move these forward more.

**Pull request merged**